File: packages/reliai-node/src/instrumentation/llamaindex.ts
import type { ReliaiClient } from "../client.ts";

const dynamicImport = new Function("specifier", "return import(specifier)") as (
  specifier: string,
) => Promise<Record<string, unknown>>;

function isFunction(value: unknown): value is (...args: unknown[]) => unknown {
  return typeof value === "function";
}

function patchMethod(
  target: Record<string, unknown>,
  methodName: string,
  wrap: (original: (...args: unknown[]) => unknown) => (...args: unknown[]) => unknown,
): void {
  const original = target[methodName];
  if (!isFunction(original) || (original as { __reliai_patched__?: boolean }).__reliai_patched__) {
    return;
  }
  const patched = wrap(original);
  (patched as { __reliai_patched__?: boolean }).__reliai_patched__ = true;
  target[methodName] = patched;
}

export async function instrumentLlamaIndex(client: ReliaiClient): Promise<void> {
  try {
    const imported = await dynamicImport("llamaindex");
    const candidates = Object.values(imported).filter((value): value is Record<string, unknown> => Boolean(value) && typeof value === "function");
    for (const candidate of candidates) {
      const prototype = candidate.prototype as Record<string, unknown> | undefined;
      if (!prototype) {
        continue;
      }
      patchMethod(prototype, "query", (original) => function patchedQuery(this: unknown, ...args: unknown[]) {
        return client.span("llm_call", () => original.apply(this, args), {
          span_type: "llm",
          framework: "llamaindex",
        });
      });
      patchMethod(prototype, "retrieve", (original) => function patchedRetrieve(this: unknown, ...args: unknown[]) {
        return client.span("retrieval", () => original.apply(this, args), {
          span_type: "retrieval",
          framework: "llamaindex",
        });
      });
    }
  } catch {
    return;
  }
}

File: packages/reliai-python/reliai/instrumentation/llamaindex.py
from __future__ import annotations

from typing import Any

from ..client import ReliaiClient


def detect_llamaindex(user_agent: str | None) -> bool:
    return bool(user_agent and "llamaindex" in user_agent.lower())


def instrument_llamaindex(client: ReliaiClient) -> None:
    for module_name in ("llama_index.core.query_engine", "llama_index.core.base.base_retriever"):
        try:
            module = __import__(module_name, fromlist=["*"])
        except Exception:
            continue
        for candidate in vars(module).values():
            if not isinstance(candidate, type):
                continue
            _patch_method(candidate, "query", client, "llm_call", {"span_type": "llm", "framework": "llamaindex"})
            _patch_method(candidate, "retrieve", client, "retrieval", {"span_type": "retrieval", "framework": "llamaindex"})


def _patch_method(cls: type[Any], method_name: str, client: ReliaiClient, span_name: str, metadata: dict[str, Any]) -> None:
    original = getattr(cls, method_name, None)
    if not callable(original) or getattr(original, "__reliai_patched__", False):
        return

    def wrapped(self: Any, *args: Any, **kwargs: Any):
        with client.span(span_name, dict(metadata)):
            return original(self, *args, **kwargs)

    setattr(wrapped, "__reliai_patched__", True)
    setattr(cls, method_name, wrapped)
